We want to allow the sharing of /dev from the host into the container.

rhatdan opened the pull request:
docker run -v /dev:/dev into the container is not acceptable solution,
since directories and devices like /dev/ptx get mishandled.

We want to be able to run libvirtd for launching vms and it needs
access to the hosts /dev.  This is a key componant of OpenStack.

Docker-DCO-1.1-Signed-off-by: Dan Walsh dwalsh@redhat.com (github: rhatdan)

crosbymichael commented:
What do you think about using `--device *` for this?

jfrazelle commented:
oooo i like the look of `--device *` also I do like this in terms of sound devices etc etc

SvenDowideit commented on docs/man/docker-create.1.md:
```diff
@@ -66,6 +67,12 @@ IMAGE [COMMAND] [ARG...]
    CPUs in which to allow execution (0-3, 0,1)
 
+**--dev**=host
+   Use the systems /dev inside of the container
```
`system's`?

SvenDowideit commented on docs/man/docker-create.1.md:
```diff
@@ -66,6 +67,12 @@ IMAGE [COMMAND] [ARG...]
+**--dev**=host
+   Use the systems /dev inside of the container
+     **host**: use the host's /dev directory inside the container.
+     No devices will be mounted on the /dev or any of its subdirs, from inside of the container
```
Does this mean to say that the container cannot mount new devices in the `/dev` directory space? or something else - I'm having trouble understanding what this sentence is trying to say

rhatdan replied:
What I was poorly trying to say was --dev=host, would also indicate that /dev/shm and /dev/mqueue would also be shared from host.  

docker currently mounts /dev, /dev/shm, and /dev/mqueue with tmpfs I believe.

rhatdan commented:
@crosbymichael I don't think that is the same thing, since if I restart the container I don't want the devices disappearing and I want other processes on the host able to see the devices I create.

crosbymichael commented:
How would you implement this then?

SvenDowideit commented:
mmm, more copy editing later, but `yes please` from me

rhatdan commented:
@crosbymichael I was hoping you would. :^)

I would guess you would bind mount the systems /dev rather then initializing the /dev.   Then don't mount any other devices on /dev.

crosbymichael commented:
So a readonly version of the host's `/dev`?

rhatdan commented:
No this would totally be shared.  Basically is is the same as -v /dev:/dev:w, but done immediately after creating the mnt namespace, before any other use of /dev.

crosbymichael commented:
So the thing causing problems is /dev/fd,ptmx, and console?  Wouldn't this mess up either the host or the container?

rhatdan commented:
I don't know, I am not sure what is going on there.  What would happen if these are shared with the hosts or other containers running with --dev=host?

jamtur01 commented on docs/sources/reference/run.md:
```diff
@@ -133,6 +133,25 @@ While not strictly a means of identifying a container, you can specify a version
+## DEV Settings
+    --dev=""  : Tell the system whether to isolate /dev or not in the container,
+           'host': use the host's /dev inside the container
+By default, all containers have their own isolated /dev that does not
```
`\dev\`

jamtur01 commented on docs/sources/reference/run.md:
```diff
@@ -133,6 +133,25 @@ While not strictly a means of identifying a container, you can specify a version
+    --dev=""  : Tell the system whether to isolate /dev or not in the container,
+           'host': use the host's /dev inside the container
+By default, all containers have their own isolated /dev that does not
+expose physical devices, only virtual devices such as /dev/null.
```
\`/dev/null\'

jamtur01 commented on docs/sources/reference/run.md:
```diff
@@ -133,6 +133,25 @@ While not strictly a means of identifying a container, you can specify a version
+By default, all containers have their own isolated /dev that does not
+expose physical devices, only virtual devices such as /dev/null.
+
+In most cases we do not want to share the hosts /dev within the container,
```
\`/dev\`

jamtur01 commented on docs/sources/reference/run.md:
```diff
@@ -133,6 +133,25 @@ While not strictly a means of identifying a container, you can specify a version
+we do this for security and to isolate the devices of the host from
+other containers.
+
+In certain cases you want your container to share the host's /dev,
```
\`/dev\`

jamtur01 commented on docs/sources/reference/run.md:
```diff
@@ -133,6 +133,25 @@ While not strictly a means of identifying a container, you can specify a version
+In certain cases you want your container to share the host's /dev,
+basically allowing processes within the container to see all of the devices
+on the system.  For example, you could build a container that uses libvirt to
+launch vms/qemu processes.  libvirt needs access to iscsi devices and can create
```
VMS/QEMU

rhatdan replied:
Should this be

VMs/QEMU

jamtur01 commented on docs/sources/reference/run.md:
```diff
@@ -133,6 +133,25 @@ While not strictly a means of identifying a container, you can specify a version
+on the system.  For example, you could build a container that uses libvirt to
+launch vms/qemu processes.  libvirt needs access to iscsi devices and can create
+network tunnel devices.  If libvirt is restarted we want these devices to remain
+and be accessible for the qemu processes, as well as the restarted libvirt container.
```
QEMU

SvenDowideit commented on docs/man/docker-create.1.md:
```diff
@@ -67,6 +68,12 @@ IMAGE [COMMAND] [ARG...]
+   Use the system's /dev inside of the container
+     **host**: use the host's /dev directory inside the container.
+     No devices will be mounted on the /dev or any of its subdirs, from inside of the container
+     Note: the host mode gives the container full access to host devices and is therefore considered insecure.
```
how full, and how insecure? is the container able to create and modify devices? `No devices will be mounted ...` line still leaves me with the feeling that `/dev` is mounted into the container in some magical read-only way, but `full access` feels like it contradicts that.

rhatdan replied:
Yes this would be very insecure and require --privileged to work.  The idea is to allow a Super Privileged Container to manipulate the host image.  For this use case we are looking at docker images as a installation medium and not really for container separation.

jfrazelle commented:
I kinda like the `--device *` because what if I wanted to do `--device /dev/snd/*` which would be awesome and then I could actually use `--device` instead of `-v` and `--privileged`

rhatdan commented:
That is fine but it is a separate issue.  In one case I want to modify the /dev that all of the processes on the host system will see.  

In your case you just want to allow a specification of additional devices available to the container.

sdake commented:
Note I'm the one that asked @rhatdan  about this feature.  He came up with the idea.  Essentially the feature comes from the kolla project upstream (http://github.com/launchpad/kolla) which has the purpose of delivering the Full OpenStack solution in containers.  The reason is because docker images offer atomic upgrades, rather then package-based upgrades.  Please see my blog here for more details on the work in question and why it is valuable for everyone in docker-land not just me:) http://sdake.io/2015/01/28/an-atomic-upgrade-process-for-openstack-compute-nodes/

libvirt spits out errors if /dev is shared with the host operating system because /dev/pts/ptmx is not readable.  I am not really sure why its not readable but Dan seemed to think this solution would solve the problem.  I'm open to other solutions as well.  Also the Kolla community wants to create devices in /dev via OpenStack Cinder (block storage system) as well as Ceph and want them to persist a recycle of the container.

I expect to completely take over the host with this use case and run with super-privileged container access.  I expect other people that export dev with this use case will want their applications to completely take over the host as well.  This is not a use case for isolation; its a use case for atomic upgrade.  The system would be secured using other mechanisms such as SELinux or AppArmor.

@crosbymichael regarding would the host and container somehow interact in a negative way, I suppose it is possible, but only testing a real live solution would show results.  If @rhatdan  has a test patch, I'll be happy to experiment with it.

sdake commented:
@rhatdan Thinking about sharing logically, the base OS already allows applications to share the /dev filesystem.  Every time you run an application that accesses /dev, that is essentially what a container does.  Whether the container sharing would result in some crash because multiple containers are trying to control the pty device, that is possible already in a bare-metal system.  The solution to that problem is the ptmx device IIUC (which serializes access to pty devices[1]).  Even if it were a device that you wouldn't normally share writes to (such as /dev/mem) - if your writing to /dev/mem, bad things are going to happen anyway :)

I'm not sure about other files besides /dev/pty/ptmx, but I suspect they are similarly protected when it makes sense for multiple users to access them (such as the pty system).

Regards
-steve

[1] http://linux.die.net/man/4/ptmx

rhatdan commented:
Thanks Steven, no test patch yet.  

I have been waiting for an updated libcontainer to work on a patch.

@crosbymichael is supposed to send me some patches that we fooled around with last week.  We were able to get a test build of libcontainer working with a shared /dev (Really the /dev was bind mounted rather then created as a tmpfs and then we commented out the other handling of /dev)  Everything seemed to work correctly.

sdake commented:
@rhatdan Great movement on this patch then!  Thanks for keeping me in the loop.  I thought this patch was stalled out because of security concerns.  While security is a concern when using SPC features, ideally the user knows what they are getting into when they use SPC features in their work.

rhatdan commented:
Right now the SPC or --privileged case is to say root within the container is the same as root outside, so don't expect these containers to contain.

icecrime commented:
@rhatdan Tentatively closing this as #11353 should solve it, but please let us know if we're wrong :-)

rhatdan commented:
This looks good to me.

mrunalp commented:
@icecrime @rhatdan Actually #11353 won't solve this one. We will need a flag to make it use bind mounts in libcontainer and changes to set that flag.

rhatdan commented:
@icecrime Looks like we need to reopen the pull request.

rhatdan commented:
@crosbymichael You never sent the patches.  I can start working on a fix for libcontainer if you have not moved forward on it.

mrunalp commented:
https://github.com/docker/libcontainer/pull/455 for the libcontainer changes.

mrunalp commented:
@rhatdan I think the libcontainer update and default_template change would have to come in an 'update libcontainer' PR once it is merged.

rhatdan commented:
@mrunalp Yes I agree, but tough to test without both.

rhatdan committed:
We want to allow the sharing of /dev from the host into the
 container.

docker run -v /dev:/dev should stop mounting other default mounts in i
libcontainer otherwise directories and devices like /dev/ptx get mishandled.

We want to be able to run libvirtd for launching vms and it needs
access to the hosts /dev.  This is a key componant of OpenStack.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

rhatdan commented:
Pushed as separate patches for now.  Once libcontainer gets merged and merged into docker, then my patch should be ready.

LK4D4 commented on daemon/execdriver/native/create.go:
```diff
@@ -219,8 +219,12 @@ func (d *driver) setupMounts(container *configs.Config, c *execdriver.Command) e
+	_, mountDev := userMounts["/dev"]
 	for _, m := range container.Mounts {
 		if _, ok := userMounts[m.Destination]; !ok {
+			if mountDev && (m.Destination == "/dev" || strings.HasPrefix(m.Destination, "/dev/")) {
```
I think it can be just `mountDev && strings.HasPrefix(m.Destination, "/dev/")`, because here `mountDev && m.Destination == "/dev"` is always false, because of `!ok`.
So if `mountDev == false` then we just skip this condition, if `mountDev == true`, then `m.Destination == "/dev" is always false.

rhatdan replied:
Yes I agree.  Thanks for explaining it.

mrunalp commented:
This PR carries the libcontainer changes https://github.com/docker/docker/pull/11502

rhatdan commented:
Yes, only way to allow people to test.

mrunalp commented:
@rhatdan #11502 is merged

rhatdan commented:
Removed the patch.

crosbymichael commented:
LGTM

ping @LK4D4

LK4D4 commented:
LGTM

rhatdan commented:
Excellent
